### DISCUSSION TITLE
Lacking way to order specific providers _from 3rd party_ libraries causing a big head ache

### DISCUSSION THREAD

**Discussion by torian257x** [laravel/framework#41334] [Ideas] [answered]
A framework should make things easy to overwrite.

statamic overwrites the 'translator' implementation e.g. `app('translator')`

Now I work on a package that wants to overwrite that translator as well. Problem is, the way it gets overwritten by statamic, does not forward calls to an existing translator like a decorator. 

furthermore, looking at the statamic translator implementation, using a decorator pattern would not even be enough for them, they need more than that e.g. maybe solve with calls to magic methods... but that is a different issue

my issue is that the providers are loaded in an order I cannot control

I tried using deferred provider, still gives me the wrong one.

Implementation details of statamic:

Provider order, they specify the provider in their own https://github.com/statamic/cms/blob/3.2/src/Providers/CpServiceProvider.php#L51

which they load here https://github.com/statamic/cms/blob/3.2/src/Providers/StatamicServiceProvider.php#L31

which they load via composer

https://github.com/statamic/cms/blob/3.2/composer.json#L58

in bootstrap/cache/packages.php

you can see that the order is alphabetically and since s for statamic/cms is very far "down" in the alphabet, it almost always gets loaded at the end I assume.

What I want is a way to say "if a provider with class name Foo\Bar::class exists, load it before my provider MyCool\Provider::class" or a variation of that.

**Comment by mikeydevelops** [comment, marked as answer]
I do not think any external package for Laravel should be able to modify its bootstrap behavior.

A compromise for you is that, if you decide to publish your package, you must make the user, add your service provider manually to `config/app.php` providers array, since any providers specified there are loaded after any package providers.

This is how Laravel bootstraps the providers:
1. Providers are collected from the `providers` array in `config/app.php`.
2. Laravel then splits the array in two parts: Framework providers, User Defined Providers.
3. Third party packages are then loaded from a package manifest previously generated from the `package:discover` command.
4. The packages are then merged in the final providers array in the following order: Framework Providers, Third Party Providers and then User Defined Providers.

**Reply by torian257x** [reply]
thanks for the explanation. This is not very transparent though. It is not very transparent how I can say "don't auto discover this package but let me manually add it to providers" either. 
I saw this: https://laravel.com/docs/9.x/packages#opting-out-of-package-discovery e.g. is that the "proper" way? e.g. don't auto discover the 3rd party package in my laravel project, then manually add it?

**Reply by mikeydevelops** [reply]
If you don't want your package auto discovered, remove your service provider from`extra->laravel->providers` array in `composer.json` of your package.

This will stop laravel from finding your package and as you said add it manually to the  providers array in `config/app.php` file.

This is how in the old days of Laravel 5 and before, users who installed packages had to add them manually to the config.
